StenosisMeasurement2D/StenosisMeasurement2D.py

```diff
@@ -0,0 +1,750 @@
+    # (in the selected parameter node).
+    self.ui.inputSliceNodeSelector.connect("currentNodeChanged(vtkMRMLNode*)", self.updateParameterNodeFromGUI)
+    self.ui.inputFiducialSelector.connect("currentNodeChanged(vtkMRMLNode*)", self.updateParameterNodeFromGUI)
+    self.ui.inputSegmentSelector.connect("currentNodeChanged(vtkMRMLNode*)", self.updateParameterNodeFromGUI)
+    self.ui.applyToAllSegmentsCheckBox.connect("toggled(bool)", self.updateParameterNodeFromGUI)
+    self.ui.limitToClosestIslandCheckBox.connect("toggled(bool)", self.updateParameterNodeFromGUI)
+
+    # Buttons
+    self.ui.applyButton.connect('clicked(bool)', self.onApplyButton)
+
+    # Make sure parameter node is initialized (needed for module reload)
+    self.initializeParameterNode()
+    
+    # Application connections
+    self.ui.inputFiducialSelector.connect("currentNodeChanged(vtkMRMLNode*)", self.observeFiducialNode)
+    
+    # Defaults
+    self.ui.optionsCollapsibleButton.collapsed = True
+    # Timer : multiple events are triggered when control point is clicked.
+    self.timer.setInterval(300)
+    self.timer.setSingleShot(True)
+    self.timer.connect("timeout()", self.onJumpToPointSingleShot)
+    # Prepare table
+    outputTable = self.ui.outputTableWidget
+    outputTable.setColumnCount(3)
+    columnLabels = ("Control point", "Segment", "Surface area")
+    outputTable.setHorizontalHeaderLabels(columnLabels)
+    # Table context menu
+    outputTable.setContextMenuPolicy(qt.Qt.CustomContextMenu)
+    outputTable.connect("customContextMenuRequested(QPoint)", self.showTableMenu)
+    """
+    Options context menu.
+    Using self.parent as the menu's parent allows the menu items to lay in the
+    foreground. Same for the table menu.
+    """
+    self.optionsMenu.setParent(self.parent)
+    optionsCollapsibleButton = self.ui.optionsCollapsibleButton
+    optionsCollapsibleButton.setContextMenuPolicy(qt.Qt.CustomContextMenu)
+    optionsCollapsibleButton.connect("customContextMenuRequested(QPoint)", self.showOptionsMenu)
+
+  def showTableMenu(self, qpoint):
+    # Start from zero.
+    self.tableMenu.clear()
+    
+    """
+    If we do this in setup, when we first enter the module, a weird background
+    rectangle is seen from the left of the module window.
+    Just workaround it.
+    """
+    self.tableMenu.setParent(self.parent)
+    
+    # Simple menu item to remove a single row.
+    actionRemoveRow = self.tableMenu.addAction("Remove row")
+    actionRemoveRow.setData(OUTPUT_TABLE_MENU_REMOVE_ROW)
+    actionRemoveRow.connect("triggered()", self.onTableMenuItem)
+    
+    # Simple menu item to remove all rows.
+    actionEmptyTable = self.tableMenu.addAction("Empty table")
+    actionEmptyTable.setData(OUTPUT_TABLE_MENU_EMPTY_TABLE)
+    actionEmptyTable.connect("triggered()", self.onTableMenuItem)
+    
+    self.tableMenu.addSeparator()
+    # Check menu item to show hide a model of the output polydata.
```

Comment:
For this it would be simpler to add a vtkBitArray column to the table, which appears as a checkbox. The checkbox is always visible and users can quickly toggle it with a single click.

![image](https://user-images.githubusercontent.com/307929/149531672-2e650ed9-aecf-4910-8498-084ba3395b82.png)

If you add a "Delete selected rows" as a separate button then you can remove the menu completely, making your code simpler.

Reply:
This one is embarrassing . I found two options.

1. Use a simple boolean value for the column. The table widget shows a checkbox, but it is always left aligned. It's OK functionally, simpler to code, but visually disturbing.

2. Add a custom widget to each cell of the column. This widget has a layout where a checkbox is centered. It's more satisfactory visually. At a functional cost. The checkbox does not trigger anything in the table when toggled, namely, it does not select a row. Connecting a lambda function to the checkbox 'toogled()' signal results in the function always working in the last row. The most acceptable solution is to hide the checkbox by default, connect the table to its 'currentCellChanged()' signal. When the user selects a cell, we know the previous and current cells, and can show/hide the checkbox. It's more complex code, with a visual drawback nevertheless : the checkbox should only appear in the current row. Else, toggling a checkbox in another row does not change current row, and operation occurs in an unwanted object.

I opted for option 2. But can revert to option 1 if you think we should prefer code simplicity namely.

Reply:
I agree that a center alignment would look nicer. That can be fixed in the table view in Slicer core. Could you take a look at it?

Reply:
> That can be fixed in the table view in Slicer core. Could you take a look at it?

There's perhaps some misunderstanding here. I'm not using a table viewable in the 'Tables' module, nor under the right side views. It's rather a qt.QTableWidget() directly, place in the module area itself. Centering the default checkbox in a cell should be done in Qt itself, AFAIU. There are many internet links where developers complain of this problem. A table cell can align strings only, and nothing else unfortunately.

Reply:
I would try to avoid creating a table in Python. It is just too much bookkeeping to make it all work correctly. It is simpler to create a vtkMRMLTableNode and add a MRML table view to see its content. If you set the table column to a vtkBitArray then it appears as a checkbox. Using the  MRML table has the advantage that its content is preserved when you save the scene and its content is also available as a CSV file when you save the scene. What do to show in that table? Can you attach a screenshot?

The checkable item's checkbox is indeed always aligned to the left, which is really, really annoying, because the checkbox works so well other than this issue. It takes a lot of effort to create a checkbox that is centered in a table cell.

Reply:
> What do to show in that table? Can you attach a screenshot?
![StenosisMeasurement2D_0](https://user-images.githubusercontent.com/33979252/149671300-c7ed9868-18bd-40a5-9f5e-ebfc3317d98f.png)

There is no real need to save the table's content in the scene as input parameters and options are saved with the scene. Applying them again is very fast. A MRML table view would upset the UI much, as all slice and 3D views will be resized, and the user will have to select again his preferred layout when switching to another module. This won't resolve the checkbox alignment problem indeed. So it's preferable to keep a QTableWidget in the module area itself.

A vtkMRMLTableNode and a MRML table view are of interest when there are large amount of data from which we plan to plot graphs, like in 'Line profile' or 'Cross-section analysis' modules, or that we may need to export for external statistical analysis. Here, a table is a convenient way to show a result. In practice, this result is rather transient in nature, and does not need long term storage, the more so that one click will produce it again.

Reply:
The control point, segment, surface area measurements could be useful to be saved automatically as a CSV file but I agree that for the model visibility and segment visibility columns MRML table would not be very useful (because it is just needed for visualization, not that important outside Slicer) and probably it would not simplify much (because you would need to synchronize the vtkBitArray items with associated model and segment visibility).

What would be useful and simplify your code a lot if there was a table for displaying/editing markups control point measurements. Because this table here is just that: a table to display values that are associated with control points. (We call these arrays "measurements" but these are often not computed automatically, but they just store static values.) 

Unfortunately, we don't have the markup control point measurement widget yet, so going with a custom table makes sense.